rules/validator/validator.go

```diff
@@ -85,147 +97,299 @@ func (v *validator) validateNamespace(ns string) error {
+		numRollupOps                  int
+		numPipelineOps                = pipeline.Len()
+	)
+	for i := 0; i < numPipelineOps; i++ {
+		pipelineOp := pipeline.At(i)
+		switch pipelineOp.Type {
+		case mpipeline.AggregationOpType:
+			numAggregationOps++
+			if numAggregationOps > 1 {
+				return errMoreThanOneAggregationOpInPipeline
+			}
+			if i != 0 {
+				return errAggregationOpNotFirstInPipeline
+			}
+			if err := v.validateAggregationOp(pipelineOp.Aggregation, types); err != nil {
+				return fmt.Errorf("invalid aggregation operation at index %d: %v", i, err)
+			}
+		case mpipeline.TransformationOpType:
+			transformOp := pipelineOp.Transformation
+			if transformOp.Type.IsBinaryTransform() {
+				transformationDerivativeOrder++
+				if transformationDerivativeOrder > v.opts.MaxTransformationDerivativeOrder() {
+					return fmt.Errorf("transformation derivative order is %d higher than supported %d", transformationDerivativeOrder, v.opts.MaxTransformationDerivativeOrder())
+				}
+			}
+			if err := validateTransformationOp(transformOp); err != nil {
+				return fmt.Errorf("invalid transformation operation at index %d: %v", i, err)
+			}
+		case mpipeline.RollupOpType:
+			// We only care about the derivative order of transformation operations in between
+			// two consecutive rollup operations and as such we reset the derivative order when
+			// encountering a rollup operation.
+			transformationDerivativeOrder = 0
+			numRollupOps++
+			if numRollupOps > v.opts.MaxRollupLevels() {
+				return fmt.Errorf("number of rollup levels is %d higher than supported %d", numRollupOps, v.opts.MaxRollupLevels())
+			}
+			if err := v.validateRollupOp(pipelineOp.Rollup, i, types); err != nil {
+				return fmt.Errorf("invalid rollup operation at index %d: %v", i, err)
+			}
+		default:
+			return fmt.Errorf("operation at index %d has invalid type: %v", i, pipelineOp.Type)
+		}
+	}
+	if numRollupOps == 0 {
+		return errNoRollupOpInPipeline
+	}
+	return nil
+}
+
+func (v *validator) validateAggregationOp(
+	aggregationOp mpipeline.AggregationOp,
+	types []metric.Type,
+) error {
+	aggregationID, err := aggregation.CompressTypes(aggregationOp.Type)
+	if err != nil {
+		return err
+	}
+	return v.validateAggregationID(aggregationID, firstLevelAggregationType, types)
+}
+
+func validateTransformationOp(transformationOp mpipeline.TransformationOp) error {
+	if !transformationOp.Type.IsValid() {
+		return fmt.Errorf("invalid transformation type: %v", transformationOp.Type)
+	}
+	return nil
+}
+
+func (v *validator) validateRollupOp(
+	rollupOp mpipeline.RollupOp,
+	opIdxInPipeline int,
+	types []metric.Type,
+) error {
+	// Validate that the rollup metric name is valid.
+	if err := v.validateRollupMetricName(rollupOp.NewName); err != nil {
+		return fmt.Errorf("invalid rollup metric name '%s': %v", rollupOp.NewName, err)
+	}
+
+	// Validate that the rollup tags are valid.
+	if err := v.validateRollupTags(rollupOp.Tags); err != nil {
```

Comment:
Yeah good question. We don't store the intermediate rollup metrics since those are just artifacts of the rollup pipeline. So IMO the debugging process probably goes like this:
* Look at metrics to identify possible causes, e.g., we currently emit metrics when a forwarded metric arrives too late at the destination server, or encounters a write error due to timeout etc. We could also log these errors with a small sample rate to control the noise level of the logs.
* For more detailed debugging, you'll need to figure out what the intermediate rollup metric IDs look like and the source and target servers, and do ngrep like we currently do for normal metrics.
* In the future we could add a `/watch` endpoint to the HTTP server in the aggregator passing in the set of IDs we want to watch, and since we know what the intermediate metric IDs look like, we can determine the servers they belong to, and we can curl those servers to live stream diagnostic information out when the watched IDs come into the server.

Reply:
Yeah the sampled logging would def help the debugging here, is there any easy way to know how does the rollup ID looks like? Maybe we should build a cmd tool for that.

Reply:
Yeah we could experiment with the sampled logging to tune the noise level.

For rollup ID, we could simply add a `match` command to `r2admin` passing in a service name and a metric ID, and the command output includes the list of (intermediate and final) rollup IDs and possibly which servers these rollup IDs belong to based on the current agg tier placement.